tests/cpp/tests/test_nodes_indexing.cpp

```diff
@@ -874,160 +892,126 @@ TEST_CASE("AdvancedIndexingNode") {
+            graph.emplace_node<ArrayValidationNode>(adv_ptr);
+
             THEN("We get the shape we expect") {
-                CHECK(adv->dynamic());
-                CHECK(std::ranges::equal(adv->shape(), std::vector{-1, 3}));
+                CHECK(adv_ptr->dynamic());
+                CHECK(std::ranges::equal(adv_ptr->shape(), std::vector{-1, 3}));
             }
 
             AND_WHEN("We create a state") {
-                auto state = graph.initialize_state();
+                auto state = graph.empty_state();
 
-                THEN("The state starts empty") {
-                    CHECK(std::ranges::equal(adv->view(state), std::vector<double>{}));
-                }
+                // as promised, but 120 values into the array initially
+                std::vector<double> values(2 * 3 * 5 * 4);
+                std::iota(values.begin(), values.end(), 0);
+                arr_ptr->initialize_state(state, values);
+
+                // dyn array starts empty
+                dyn_ptr->initialize_state(state, std::vector<double>{});
+
+                graph.initialize_state(state);
 
                 AND_WHEN("We mutate the main array, grow the indexing nodes and propagate") {
-                    std::vector<double> values(2 * 3 * 5 * 4);
-                    std::iota(values.begin(), values.end(), 0);
+                    // double the size of the array
                     arr_ptr->grow(state, values);
-                    // i_ptr -> {0, 1}
-                    // j_ptr -> {1, 2}
-                    // k_ptr -> {3, 3}
-                    dyn_ptr->grow(state, {0, 1, 3, 1, 2, 3});
+                    dyn_ptr->grow(state);
 
-                    arr_ptr->propagate(state);
-                    dyn_ptr->propagate(state);
-                    i_ptr->propagate(state);
-                    j_ptr->propagate(state);
-                    k_ptr->propagate(state);
-                    adv->propagate(state);
-
-                    THEN("The state has the expected values and the diff is correct") {
-                        CHECK(adv->size(state) == 6);
-                        CHECK(std::ranges::equal(adv->view(state),
-                                                 std::vector{7, 27, 47, 71, 91, 111}));
-                        verify_array_diff({}, {7, 27, 47, 71, 91, 111}, adv->diff(state));
-                    }
+                    graph.propagate(state, graph.descendants(state, {arr_ptr, dyn_ptr}));
 
-                    AND_WHEN("We mutate the main array again") {
-                        arr_ptr->commit(state);
-                        dyn_ptr->commit(state);
-                        i_ptr->commit(state);
-                        j_ptr->commit(state);
-                        k_ptr->commit(state);
-                        adv->commit(state);
+                    AND_WHEN("We commit") {
+                        graph.commit(state, graph.descendants(state, {arr_ptr, dyn_ptr}));
 
-                        arr_ptr->set(state, 47, -1);
```

Comment:
Why are we removing the tests that mutate the main array and check for correctness?

Reply:
We do mutate both, see line 941 above.

Reply:
We do also check for correctness, but mostly rely on the `ArrayValidationNode` for consistency.

Reply:
You grow the array, but what about updates standard updates without changing size?

Reply:
Oh, I see. Sure. I can restore that.

Reply:
Added back in in 5b2c1ee